library/alloc/src/vec/source_iter_marker.rs

```diff
@@ -1,3 +1,100 @@
+///! 2. loop through the iterator, for each step the source is required to have advanced
+///!    its read pointer at least once so that the result can be written into the allocation
+///!    without overwriting yet-to-be-iterated items. Also counts the number of items written.
+///! 3. on panic the written items must be dropped before the iterator is dropped which will
+///!    then drop any unread items. This requires the write count to be accurate when the panic occurs.
+///! 4. on success the source is accessed again, any unread items are dropped and then the allocation
+///!    is stolen so it doesn't get dropped with the iterator.
+///!    Since this changes the state of the source while still being owned by the iterator adapters
+///!    they must be able to handle this situation in their drop impl.
+///!    This is a safety requirement of `SourceIter`.
+///! 5. assemble a new vec from From the capacity, count of written items and the allocation pointer
+///!
+///!
+///! Additionally step 2 is further specialized when the iterator implements
+///! `TrustedRandomAccessNoCoerce` to let the optimizer see that it is a counted loop with a single
+///! induction variable. This can turn some iterators into a noop, i.e. it reduces them from O(n) to
+///! O(1). Since unchecked accesses through that trait do not advance the read pointer of `IntoIter`
+///! this would interact unsoundly with steps 3 and 4. But since the normal `Drop` implementation of
+///! `IntoIter` would suffer from the same problem it is only correct for `TrustedRandomAccessNoCoerce`
+///! to be implemented when the items don't have a destructor. Thus that implicit requirement
+///! also makes the specialization safe to use for in-place collection.
+///!
+///! The combined requirements of all these unsafe traits and the need to implement them on each
+///! part of the iterator pipeline - since there is no central struct evaluating an iterator
+///! holistically - leads to the need for non-local reasoning, i.e. the safety has to be evaluated
+///! taking the behavior of `IntoIter`, the adapters and this specialization into account.
+///! Interactions with other features such as clonable iterators, panics inside user-provided closures
+///! and partially consumed (including reverse iteration) ones also have to be considered.
+///! For example a previous implementation of `InPlaceIterable` on `Peek` turned out to be unsound
+///! due to `IntoIter`'s `Clone` implementation shortening the underlying allocation while retaining
+///! the extra peeked item which then causes the iterator to yield more items than the allocation can
+///! hold.
+///!
+///!
+///! Some examples that are optimized by this specialization, more can be found in the `Vec`
+///! benchmarks:
+///!
+///! ```rust
+///! /// Converts a usize vec into an isize one.
+///! fn cast(vec: Vec<usize>) -> Vec<isize> {
+///!   // Does not allocate, free or panic. On optlevel>=2 it does not loop.
```

Comment:
The "On optlevel>=2 it does not loop" is unfortunately true only for some cases. Sometimes LLVM produces empty loops.

Reply:
Well, it's true for the usize/isize case, there even is a codegen test for that. And that it refers to opt-level itself should make clear that it depends on optimizations, no?

Reply:
My observation was meant to be something that might be worth looking into (and maybe highlight in the docs) rather than a correction.

Reply:
Oh, yeah, there already is #79308 for that. I can link to it.